### PR TITLE
(Preview) Fix no same naming convention between template and form

### LINKED ISSUE #22833: Holding the values in form when the post request is failed with error
### Description

At first, I found that some vaules in the form were missing or changed to default values automaticly, when creating an org or a repo with invalid name.
e.g.
image

`Visibility` in this page is required so users will be aware of the change.
But if there are some no required params with this bug, user may post wrong data with no awareness of these changes.

Then I found func `AssignForm` in modules/web/middleware/binding.go which is used to assign form values back to the template data.
It seems we need to define same variable names between form data and template data?

e.g.
`Visibility` in this page is defined as `DefaultOrgVisibilityMode` in template data, but `visibility` in form data.

### Gitea Version

87261f3fb95da92d74d72d057fc0f1e9819f16a7

### Can you reproduce the bug on the Gitea demo site?

No

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

build

### Database

None

### PR DESCRIPTION
**Pull request by yp05327**
Fixes https://github.com/go-gitea/gitea/issues/22833

(This PR is still on working)

The variable's names in form data will be changed to snakecase in func `AssignForm` if `form` tag is not defined.
And this makes contributors write various code which follows different naming convention of variables.
Then low readability code caused the bug I mentioned in https://github.com/go-gitea/gitea/issues/22833
(I personally think so)

So I made a change to `AssignForm`:
The variable's names in form data will be changed into field's name.
Then we just let the variable's names in context data as same as the field's name which is simply defined in each form struct.
(easy to check whether the variable's name is currect, otherwise the vaule of therm will be changed when `RenderWithErr` or some unknow bugs will be occured?)

This is a big change, because we need to check all functions uses `AssignForm`.
So I made this preview PR, if it is worth to spend time on this work, I will continue to work on this.

### CODE REVIEW

File: modules/web/middleware/binding.go [hunk 1]
```diff
@@ -42,7 +42,7 @@ func AssignForm(form interface{}, data map[string]interface{}) {
 		if fieldName == "-" {
 			continue
 		} else if len(fieldName) == 0 {
-			fieldName = util.ToSnakeCase(field.Name)
+			fieldName = field.Name
```
Comment (wxiaoguang) [inline modules/web/middleware/binding.go]: Maybe the snake_case is done by purpose:

* snake_case: for auto filled form, matching the form element's  `name`
* PascalCase:  for non-form variables

?

If  the form-variables  and  non-form-variables  are mixed together, it  would  be very difficult to  understand what a variable is  used for, and may cause uncontrollable overwriting?

Comment (yp05327) [reply modules/web/middleware/binding.go]: @wxiaoguang 
Thanks for your review.
You are right, it is difficult to distinguish and may cause uncontrollable overwriting.

How about the following changes.

For non-form variables:
```
// in func
ctx.Data["HelloData"]
// in template
{{.HelloData}}
```
For auto filled form:
```
// in func
ctx.Data["Form"] = forms.FormStruct {
    Variable: vaule,
    // ....
}
// in template
{{.Form.Variable}}
```

Comment (wxiaoguang) [reply modules/web/middleware/binding.go]: I am not sure and I haven't thought about it carefully.

Maybe the Technical Oversight Committee could help.

Comment (wolfogre) [reply modules/web/middleware/binding.go]: Personally, I like the idea `{{.Form.Variable}}`, it's more readable and go-style. And there will be a clear way to check if it's an error page to repost with entered form: `{{if .Form}} /* keep the values in .Form */ {{end}}`

**Comment by wolfogre** [comment]
Since this is a big decision, I think I should call the TOC.

@lunny @techknowlogick @6543 @zeripath @jolheiser

File: modules/web/middleware/binding.go [hunk 2]
```diff
@@ -26,27 +26,8 @@ func init() {
...
-		// Allow ignored fields in the struct
-		if fieldName == "-" {
-			continue
-		} else if len(fieldName) == 0 {
-			fieldName = util.ToSnakeCase(field.Name)
-		}
-
-		data[fieldName] = val.Field(i).Interface()
-	}
+	// TODO: Allow ignored fields in the struct?
```
Comment (zeripath) [inline modules/web/middleware/binding.go:29]: Don't worry about these - that's more for the auto-binding thing to care about.

File: modules/web/middleware/binding.go (hunk 1 above)
Comment (zeripath) [reply modules/web/middleware/binding.go]: I see the point of the snake_case but using the `{{.Form}}` would prevent us from having to do that.

Comment (zeripath) [reply modules/web/middleware/binding.go]: So there is one potential issue with the `.Form` object if you're using go code.

I'm not certain if there's any code that does this - but I could imagine some code that doesn't have the type of the Form object but would have been able to update the values in the `ctx.Data["some_field"]` which by convention was a string.

**Comment by zeripath** [comment]
This is a cautious :+1: from me. The `snake_case` form mapping is weird and feels fragile.

My caution is because when you put stuff in `ctx.Data` it becomes an `interface{}` and whilst that field values were also `interface{}` the number of likely types was very small.

I think there are a few places where we reuse the same or similar templates with different forms but I'm not sure if we have go code that would adjust/use the snake_cases fields. If we do that - then the `.Form` variant could be quite difficult to do.

---

As an aside... in order to make migration safer and easier I would recommend leaving in the current auto snake_case mapping but also just stick the Form in {{.Form}} and leave a note to remove it soon.

**Comment by lunny** [comment]
> Since this is a big decision, I think I should call the TOC.
> 
> @lunny @techknowlogick @6543 @zeripath @jolheiser

An advantage of the new `Form` in Golang template is maybe it could reduce the potential risk to have the same key in the response context `Data` map. But once we decide it, I think there are many templates that need to be changed.

**Review by delvh — APPROVED** [review]
Well… Everything seems to work.
Although I haven't understood yet how the other contents are auto-injected, when all you do is swap out the maps you use…

**Comment by yp05327** [comment]
@delvh 
This pr dosen't work well now. 
As @zeripath said here https://github.com/go-gitea/gitea/pull/22846#discussion_r1103170806, there are some potential problems need to be fixed.
I'm working on https://github.com/go-gitea/gitea/pull/22865, and after finished this, I will work on this PR.

File: modules/web/middleware/binding.go (hunk 1 above)
Comment (yp05327) [reply modules/web/middleware/binding.go]: In some pages, there are several forms in one page. Need to consider a new solution to solve this problem.

Comment (yp05327) [reply modules/web/middleware/binding.go]: > In some pages, there are several forms in one page. Need to consider a new solution to solve this problem.

https://github.com/go-gitea/gitea/pull/22846/commits/a64ba42e1b679aa9699e4adee2dc71b9f5593e1f is an example to fix  this problem.

**Comment by wxiaoguang** [comment]
I have some clearer ideas about this problem now.

The root problem is that it's very difficult to re-fill the forms with correct values, the case is:

1. Open a "GET" page, fill with default modal values.
2. Do "POST", if error occurs, fill the form with posted values.

There is no clear / general approach for various cases/pages. I am pretty sure because I have made various different web frameworks by Java/PHP/Go, the "form-post-refill" problem has no clear solution.

And even worse, there are a lot of misuses in old form-binding code (aka `m.Post("", web.Bind())`), some mechanisms are just broken.

----

Then, how to make the UI/UX friendly to end users if something wrong happens? I think the answer is AJAX.

1. Open a "GET" page, fill with default modal values.
2. Do "POST" with AJAX, if error occurs, just display the error message, the user is still on the current page with all values.

And the AJAX Form can handle more cases, eg: internal server error, broken network, etc.

**Comment by yp05327** [comment]
If we use AJAX, the change will be much bigger.
But maybe it is a good change as we can combine the codes of handling POST requests from frontend and API request easily.
I found that sometimes we changed the logic of frontend but forgot to apply the changes to API, (e.g. https://github.com/go-gitea/gitea/issues/23776)
and different logics between frontend and API (e.g. https://github.com/go-gitea/gitea/issues/23934).

**Comment by wxiaoguang** [comment]
IMO it's not that big, the most work is rewriting the `XxxPost` handlers, and these handlers could be simplified a lot because they do not need to handle various errors & re-rendering , just responding the JSON is enough.

**Comment by yp05327** [comment]
Is it possible to reuse API code?
Or just let all frontend Post request to be an internal api call/direct api call in the backend?

For example:
Client: Post /xxx/xxx/xxx
↓
internal request: Post /api/v1/xxx/xxx/xxx

Or
Client: Post /api/v1/xxx/xxx/xxx

**Comment by wxiaoguang** [comment]
> Is it possible to reuse API code?

No, the API handlers and Web handlers shouldn't be mixed together.

API handlers focuses on stability, they do not change too much.

Web handlers focuses on flexibility, and are only used by Web UI (so no swagger document)

> `Client: Post /xxx/xxx/xxx -> internal request: Post /api/v1/xxx/xxx/xxx` OR `Client: Post /api/v1/xxx/xxx/xxx`

Neither. The AJAX forms could just post to the original "post" handlers, it could be a simple change: from `<form action>` to JS form submit.

**Comment by wxiaoguang** [comment]
Now there is a framework level support for making AJAX requests.

Here is a sample for how to do it:  Use fetch to send requests to create issues/comments #25258 


I think this issue can be closed. Move more forms to the AJAX world.

**Pull request closed without merging**